### PR TITLE
Feature: State

### PR DESCRIPTION
**Pull request by YehorPytomets**
### Changes introduced by this PR:
* Add `State` for keeping `FileListExplorer`'s data.
* Add `Action` and `GetFolderContentAction` to initiate retrieving data from the server.
* Add `mutators` to change `State`'s content on different stages of it's lifecycle.
* Add `StateManager` to manage access to `State`.

### CODE REVIEW

File: webclient/app/mocks/server-mock.js [hunk 1]
```diff
@@ -46,6 +46,18 @@ class ServerMock {
 
       return this._createTokenResponse();
     });
+
+    fetchMock.get('/explorer', (url, options) => {
```
Comment (dkalinin) [inline webclient/app/mocks/server-mock.js:50]: - [ ] Please use URLs that are specified on miro board.

File: webclient/app/mutators/get-list-items-mutator.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+export default class GetListItemsMutator extends Mutator {
+
+  /**
+   * Instantiates List Items mutator.
+   *
+   * @param {object[]} listItems - list of items to set to the state.
+   */
+  constructor(listItems) {
+    super({listItems});
```
Comment (dkalinin) [inline webclient/app/mutators/get-list-items-mutator.js:16]: - [ ] What's the purpose of wrapping parameters with object? We can save listItems to a private property and there's no need to call parent constructor.

File: webclient/app/mutators/get-list-items-mutator.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+   * @return {null|FolderModel|FileModel} - instantiated model.
+   */
+  _createModel(item) {
+    const listItemCreators = {
+      'file': () => new FileModel(item),
+      'folder': () => new FolderModel(item),
+    };
+    const createModel = listItemCreators[item.type];
+
+    return createModel ? createModel() : null;
```
Comment (dkalinin) [inline webclient/app/mutators/get-list-items-mutator.js:41]: - [ ] If we can't create a model then this is an error but not null.

File: webclient/app/components/state-aware-component.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import Component from './component.js';
+
+/**
+ * Component that relies on storing it's data in @see {@link State}.
+ * @abstract
+ */
+export default class StateAwareComponent extends Component {
```
Comment (dkalinin) [inline webclient/app/components/state-aware-component.js:7]: - [ ] Let's also have a dispatch method which delegates it to state manager.

File: webclient/app/mutators/mutator.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * @abstract
+ */
+export default class Mutator {
+
+  constructor(configuration) {
+    Object.assign(this, configuration);
+  }
+
+  /**
+   * Applies changes im State object.
```
Comment (dkalinin) [inline webclient/app/mutators/mutator.js:12]: - [ ] in

File: webclient/app/pages/file-list-explorer.js [hunk 6]
```diff
@@ -91,7 +108,64 @@ export default class FileListExplorer extends Component {
...
+  initState() {
+    const {listComponent} = this;
+
+    this.onStateChanged('listItems', (event) => {
+      listComponent.renderListItems(event.detail.state.listItems);
+    });
+
+    this.onStateChanged('isListItemsLoading', (event) => {
+      const loader = this.rootElement.querySelector('[data-type="loader"]');
+      const {classList} = loader;
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:132]:
```suggestion
      const classList = this.rootElement.querySelector('[data-type="loader"]').classList;
```

File: webclient/tests/app/actions/get-folder-content-action-tests.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+export default (() => {
+  const {module, test} = QUnit;
+
+  module('Get Folder Content Action');
+
+  test('should correctly initiate request for folder content.', async (assert) => {
+    assert.expect(2);
+
+    const state = new State();
+    const apiService = new ApiServiceMock(2, true);
```
Comment (dkalinin) [inline webclient/tests/app/actions/get-folder-content-action-tests.js:18]: - [ ] We should not use the same fetch-mock for development and for testing.

File: webclient/tests/app/services/api-service-tests.js [hunk 8]
```diff
@@ -132,24 +120,175 @@ export default (() => {
...
+   * @param {Qunit.assert} assert - assertions object.
+   * @param {string} url - request path.
+   * @param {object} expectedOptions - request configuration.
+   * @param {number} numberOfCalls - number of calls that was done to passed url.
+   */
+  function assertCalledRoute(assert, url, expectedOptions, numberOfCalls = 1) {
+    assert.strictEqual(fetchMock.calls(url, expectedOptions).length, numberOfCalls,
+      `Should make request to '${url}' ${numberOfCalls} ${numberOfCalls === 1 ? 'time' : 'times'} in the test.`,
+    );
+    assert.ok(fetchMock.done(url), `Should make request to '${url}' the expected number of times.`);
```
Comment (dkalinin) [inline webclient/tests/app/services/api-service-tests.js:292]: - [ ] This doesn't check expected calls number.

File: webclient/app/mocks/server-mock.js (hunk 1 above)
Comment (YehorPytomets) [reply webclient/app/mocks/server-mock.js]: I've changed route from `'/explorer'` to `'folder/:folderId/content'`. 
But, as soon as reading `folderId` is implemented in Feature: Get Folder Content I pass mock folder if from `FileListExplorer` to maintain current implementation.

File: webclient/app/mutators/get-list-items-mutator.js (hunk 2 above)
Comment (YehorPytomets) [reply webclient/app/mutators/get-list-items-mutator.js]: I've thought that using `Object.assing(this, properties)` in super constructor as one of the ways of assigning passed parameters.
For now I assign passed parameters right in called constructor.

File: webclient/tests/app/actions/get-folder-content-action-tests.js (hunk 7 above)
Comment (YehorPytomets) [reply webclient/tests/app/actions/get-folder-content-action-tests.js]: I've looked through `get-folder-content-action-tests.js` and didn't find usages of fetch-mock for development.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready for review.

File: webclient/app/pages/file-list-explorer.js (hunk 6 above)
Comment (YehorPytomets) [reply webclient/app/pages/file-list-explorer.js]: Done.

File: webclient/tests/app/actions/get-folder-content-action-tests.js [hunk 9]
```diff
@@ -15,11 +15,14 @@ export default (() => {
     assert.expect(2);
 
     const state = new State();
-    const apiService = new ApiServiceMock(2, true);
+    const apiService = new ApiServiceMock({
```
Comment (dkalinin) [inline webclient/tests/app/actions/get-folder-content-action-tests.js:18]: - [ ] Please don't use dev mock server in tests. Please replace with direct usage of fetch-mock.

File: webclient/app/mutators/get-list-items-mutator.js [hunk 10]
```diff
@@ -7,13 +7,30 @@ import FolderModel from '../models/folder-model.js';
...
+    super();
+    this._listItems = listItems;
+  }
+
+  /**
+   * Getter for list items.
+   *
+   * @return {Object[]} retrieved list items.
+   */
+  get listItems() {
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply webclient/app/mutators/get-list-items-mutator.js:32]: I used this to get access to list items when testing order of called mutators in `GetFolderContentAction` tests.
From now I use private properties to refer to list items.

File: webclient/app/services/api-service.js [hunk 11]
```diff
@@ -80,11 +80,13 @@ export default class ApiService {
   /**
    * Retrieves list items from the server.
    *
-   * @return {Promise<Response, Error>} - promise, resolved for client processing.
+   * @param {object} configuration - request configuration.
+   * @param {string} configuration.folderId - retrieve folder content by that id.
+   * @return {Promise<Response, Error>} - promised, resolved for client processing.
    */
-  async getFolderContent() {
+  async getFolderContent({folderId}) {
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply webclient/app/services/api-service.js:87]: I use such approach in `feature/get-folder-content` when passing location params object to `ApiService.getFolderContent()`. 
But for now I pass only `folderId` value

File: webclient/tests/app/actions/get-folder-content-action-tests.js (hunk 9 above)
Comment (YehorPytomets) [reply webclient/tests/app/actions/get-folder-content-action-tests.js:18]: I didn't find usage of dev mock server neither in `ApiServiceMock`, nor in `ApiService`. Dev mock get enabled in `Applicaion` component. Maybe I'm missing something, could you please explain that issue during discussion?

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

**Pull request merged**